gslib/utils/copy_helper.py

```diff
@@ -2534,6 +2534,11 @@ def __init__(self, fp, tracker_file_name, src_obj_metadata, start_byte,
     self._start_byte = start_byte
     self._end_byte = end_byte
 
+  @property
+  def mode(self):
+    """Returns the mode of the underlying file descriptor, or None."""
+    return getattr(self._orig_fp, 'mode', None)
```

Comment:
In what circumstance would the original file descriptor's mode be none? Only if we construct a `SlicedDownloadFileWrapper` object without a valid `fp`?

Asking mostly out of curiosity if this can happen in the wild, and if we should test for that circumstance, and if we're relying on the calling function to handle the error.

Reply:
I don't know of such a scenario, but `write_to_fd()` has an else clause (which was being hit in the scenario I'm fixing) that's hit when the `fd` doesn't have a `mode` attribute. Maybe this was added solely because of this `SlicedDownloadFileWrapper`, or maybe there are other wrapper classes like this too?

Come to think of it, I should probably add this property to all of our Upload/Download Wrapper classes.  To answer your question, I don't think this should ever occur in practice, but I don't know that for sure.  We should probably add a statement in `write_to_fd()` to print a debug statement whenever it writes to a file descriptor that doesn't have a `mode` attribute (although this could get pretty spammy real quick if you're writing lots of blocks to a large file...)

Reply:
Added the `mode` property to other wrapper classes as well, just pushed a commit. PTAL.

Reply:
(I think I got them all; the command I used to find these classes was `grep "class [A-Z][A-Za-z_0-9]\+Wrapper" . -r`).

Reply:
Understood! New commit LGTM! :boat: